pkg/health/health.go

```diff
@@ -0,0 +1,160 @@
+package health
+
+import (
+	"context"
+	"time"
+
+	"github.com/GoogleCloudPlatform/cloud-run-release-operator/internal/metrics"
+	"github.com/GoogleCloudPlatform/cloud-run-release-operator/internal/util"
+	"github.com/GoogleCloudPlatform/cloud-run-release-operator/pkg/config"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+// DiagnosisResult is a possible result after a diagnosis.
+type DiagnosisResult int32
+
+// Possible diagnosis results.
+const (
+	Unknown      DiagnosisResult = 0
+	Inconclusive DiagnosisResult = 1
+	Healthy      DiagnosisResult = 2
+	Unhealthy    DiagnosisResult = 3
+)
+
+// Diagnosis is the information about the health of the revision.
+type Diagnosis struct {
+	OverallResult DiagnosisResult
+	CheckResults  []*CheckResult
+}
+
+// CheckResult is information about a metrics criteria check.
+type CheckResult struct {
+	Threshold     float64
+	ActualValue   float64
+	IsCriteriaMet bool
+}
+
+// Diagnose attempts to determine the health of a revision.
+//
+// If the minimum number of requests is not met, then health cannot be
+// determined and Diagnosis.EnoughRequests is set to false.
+//
+// Otherwise, all metrics criteria are checked to determine if the revision is
+// healthy.
+func Diagnose(ctx context.Context, provider metrics.Metrics, query metrics.Query,
+	offset time.Duration, minRequests int64, healthCriteria []config.Metric) (*Diagnosis, error) {
+
+	logger := util.LoggerFromContext(ctx)
+	metricsValues, err := CollectMetrics(ctx, provider, query, offset, healthCriteria)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not collect metrics")
+	}
+
+	overallResult := Healthy
+	var results []*CheckResult
+	for i, criteria := range healthCriteria {
+		result := determineResult(criteria.Type, criteria.Threshold, metricsValues[i])
+		results = append(results, result)
+
+		if !result.IsCriteriaMet {
+			overallResult = Unhealthy
+
+			logger := logger.WithFields(logrus.Fields{
+				"metricsType": criteria.Type,
+				"threshold":   criteria.Threshold,
+				"actualValue": result.ActualValue,
+			})
+			// If criteria is a latency, we want to log the percentile as well
+			if criteria.Type == config.LatencyMetricsCheck {
+				logger = logger.WithField("percentile", criteria.Percentile)
```

Comment:
just log it always, it'll be a percentile=0 field that doesn't mean anything. it's cleaner that way (+log stms align in human-readable output)

Reply:
Fixed!